lib/companies/email_events.ex

```diff
@@ -0,0 +1,57 @@
+defmodule Companies.EmailEvents do
+  @moduledoc """
+  Tracking email lifecycle events from SendGrid
+
+  https://sendgrid.com/docs/for-developers/tracking-events/
+  """
+
+  alias Companies.Schema.{EmailEvent, User}
+  alias Companies.{Accounts, Repo, TaskSupervisor}
+
+  @disable_user_emails ["bounce", "spamreport", "unsubscribe"]
+  @whitelisted_events ["delivered", "open"]
+
+  def create(events) do
+    Task.Supervisor.start_child(TaskSupervisor, fn ->
+      Enum.each(events, &create_event/1)
+    end)
+  end
+
+  defp create_event(%{"email" => user_email, "event" => event} = params) do
+    with true <- watched_event?(event),
+         %User{} = user <- Accounts.get_by_email(user_email),
+         {:ok, _event} <- store_event(params, user),
+         {:ok, _user} <- update_user_preferences(event, user) do
+      :ok
+    else
+      false ->
+        :ok
+
+      {:error, %{errors: errors}} ->
+        send_error(errors, params)
+    end
+  end
+
+  defp send_error(errors, params) do
+    Appsignal.send_error(%RuntimeError{}, "SendGrid Callback", [], %{params: params, errors: errors})
```

Comment:
If something does go sideways in our asynchronous write, let's tell AppSignal about it.